src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,153 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use abi::{ArgType, FnType, LayoutExt, Reg, Uniform};
+use abi::{ArgAttribute, ArgType, CastTarget, FnType, LayoutExt, PassMode, Reg, RegKind, Uniform};
 use context::CodegenCx;
+use rustc::ty::layout::{self, Size};
 
-use rustc::ty::layout::Size;
+fn extend_integer_width_mips(arg: &mut ArgType, bits: u64) {
+    // This is like ArgType::extend_integer_with_to, but always sign extends
+    // integers >= 32 bits as required by the n64 ABI
+    if let layout::Abi::Scalar(ref scalar) = arg.layout.abi {
+        if let layout::Int(i, signed) = scalar.value {
+            if i.size().bits() < bits {
+                if let PassMode::Direct(ref mut attrs) = arg.mode {
+                    attrs.set(if signed || i.size().bits() >= 32 {
+                        ArgAttribute::SExt
+                    } else {
+                        ArgAttribute::ZExt
+                    });
+                }
+            }
+        }
+    }
+}
+
+fn bits_to_int_reg(bits: u64) -> Reg {
+    if bits <= 8 {
+        Reg::i8()
+    } else if bits <= 16 {
+        Reg::i16()
+    } else if bits <= 32 {
+        Reg::i32()
+    } else {
+        Reg::i64()
+    }
+}
 
-fn classify_ret_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>,
-                             ret: &mut ArgType<'tcx>,
-                             offset: &mut Size) {
+fn classify_ret_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, ret: &mut ArgType<'tcx>) {
     if !ret.layout.is_aggregate() {
-        ret.extend_integer_width_to(64);
+        extend_integer_width_mips(ret, 64);
+        return;
+    }
+
+    let size = ret.layout.size;
+    let bits = size.bits();
+    if bits <= 128 {
+        // Unlike other architectures which return aggregates in registers, MIPS n64 limits the
+        // use of float registers to structures (not unions) containing exactly one or two
+        // float fields.
+
+        if let layout::FieldPlacement::Arbitrary { .. } = ret.layout.fields {
+            let float_regs: Vec<Reg> =
+                (0..ret.layout.fields.count()).flat_map(|i| {
+                    match ret.layout.field(cx, i).abi {
+                        layout::Abi::Scalar(ref scalar) => match scalar.value {
+                            layout::F32 => Some(Reg::f32()),
+                            layout::F64 => Some(Reg::f64()),
+                            _ => None
+                        },
+                        _ => None
+                    }
+                }).collect();
+
+            if ret.layout.fields.count() == float_regs.len() {
+                if float_regs.len() == 1 {
+                    ret.cast_to(Uniform { unit: float_regs[0], total: float_regs[0].size });
```

Comment:
We have conversion impls so you can specifically do `ret.cast_to(float_regs[0])`.

Reply:
Fixed.